kb/how-to/add-test-generation.md

```diff
@@ -0,0 +1,478 @@
+null strings.
+For calculating the expected result, these inputs are also converted to, and stored in,
+`int` form; this is a multi-precision type in Python and so can be used for bignum
+calculations.
+
+For the example, add the following constructor to `BignumAddExample`:
+
+```python
+    def __init__(self, val_a: str, val_b: str) -> None:
+        self.arg_a = val_a
+        self.arg_b = val_b
+        # int(val, 16) converts from a hex string to int
+        # empty strings should be treated as zero in calculations
+        self.int_a = int(val_a, 16) if val_a else 0
+        self.int_b = int(val_b, 16) if val_b else 0
+```
+
+### Implementing abstract BaseTarget methods
+
+Abstract methods declared in `BaseTarget` must also be implemented in test function
+classes.
+
+For the example, the following imports will be required:
+
+```python
+import itertools
+from typing import Iterator, List
+# After import scripts_path
+from mbedtls_dev import test_case
+```
+
+`arguments()` returns the list of arguments required by the test function.
+Any C-string arguments must be quoted.
+In the example, a `result()` method is also added to calculate the expected output.
+
+For the example, add the following methods to `BignumAddExample`:
+
+```python
+    def arguments(self) -> List[str]:
+        return [
+            "\"{}\"".format(self.arg_a),
+            "\"{}\"".format(self.arg_b),
+            self.result()
+        ]
+
+    def result(self) -> str:
+        # Return as a quoted hexadecimal string, without leading 0x
+        return "\"{:x}\"".format(self.int_a + self.int_b)
+```
+
+`generate_function_tests()` handles the construction of class instances for each unique
+test case, and yields `test_case.TestCase` objects from each instance.
+This is a `classmethod`, which acts on the class, rather than a specific instance.
+In the example, a list of strings will be used as inputs, and a test case is generated
+for each possible combination.
+The standard module `itertools` is used for generation of unique input combinations.
+
+For the example, add the following method to `BignumAddExample`:
+
+```python
+    @classmethod
+    def generate_function_tests(cls) -> Iterator[test_case.TestCase]:
+        # Create a list of input values
+        input_values = ["", "0", "1", "123"]
+        # combinations_with_replacement generates a list of all unique
+        # combinations, including where inputs are the same i.e. ("0", "0")
+        for a_value, b_value in itertools.combinations_with_replacement(
+            input_values, 2
+        )
+            # Initialize a class instance with these values
+            class_instance = cls(a_value, b_value)
+            # Yield the TestCase object from this instance
+            yield class_instance.create_test_case()
+```
+
+### Other BaseTarget attributes and methods
+
+`description()` returns a description of the test case, in the format "`test_name`
+#`count` `case_description`" by default.
```

Comment:
Don't put `#` as the first character of a line: in many markdown variants, it starts a title even when the previous line is not blank.